main/src/com/google/refine/commands/recon/GuessTypesOfColumnCommand.java

```diff
@@ -170,70 +183,66 @@ protected IndividualQuery(String query, int limit) {
-                connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded; charset=UTF-8");
-                connection.setConnectTimeout(30000);
-                connection.setDoOutput(true);
-                
-                DataOutputStream dos = new DataOutputStream(connection.getOutputStream());
-                try {
-                    String body = "queries=" + ParsingUtilities.encode(queriesString);
-                    
-                    dos.writeBytes(body);
-                } finally {
-                    dos.flush();
-                    dos.close();
+            RequestConfig defaultRequestConfig = RequestConfig.custom()
+                    .setConnectTimeout(30 * 1000)
+                    .setConnectionRequestTimeout(30 * 1000)
+                    .setSocketTimeout(10 * 1000).build();
+
+            HttpClientBuilder httpClientBuilder = HttpClients.custom()
+                    .setUserAgent(RefineServlet.getUserAgent())
+                    .setRedirectStrategy(new DefaultRedirectStrategy(new String[] {
+                            HttpGet.METHOD_NAME,
+                            HttpHead.METHOD_NAME,
+                            HttpPost.METHOD_NAME }))
+                    .setDefaultRequestConfig(defaultRequestConfig);
+            
+            CloseableHttpClient httpClient = httpClientBuilder.build();
+            HttpPost request = new HttpPost(serviceUrl);
+            List<NameValuePair> body = Collections.singletonList(
+                    new BasicNameValuePair("queries", queriesString));
+            request.setEntity(new UrlEncodedFormEntity(body, Consts.UTF_8));
+            
+            try (CloseableHttpResponse response = httpClient.execute(request)) {
+                StatusLine statusLine = response.getStatusLine();
+                if (statusLine.getStatusCode() >= 400) {
+                    throw new IOException("Failed  - code:" 
+                            + Integer.toString(statusLine.getStatusCode()) 
+                            + " message: " + statusLine.getReasonPhrase());
                 }
                 
-                connection.connect();
-            }
-
-            if (connection.getResponseCode() >= 400) {
-                InputStream is = connection.getErrorStream();
-                throw new IOException("Failed  - code:" 
-                        + Integer.toString(connection.getResponseCode()) 
-                        + " message: " + is == null ? "" : ParsingUtilities.inputStreamToString(is));
-            } else {
-                InputStream is = connection.getInputStream();
-                try {
-                    String s = ParsingUtilities.inputStreamToString(is);
-                    ObjectNode o = ParsingUtilities.evaluateJsonStringToObjectNode(s);
+                String s = ParsingUtilities.inputStreamToString(response.getEntity().getContent());
+                ObjectNode o = ParsingUtilities.evaluateJsonStringToObjectNode(s);
```

Comment:
This needs to handle a JsonParseException either here or in the surrounding try/except block.

Reply:
JsonException is a sublcass of IOException, so isn't the current handling of IOException enough? What specific action do you want to take when a JsonParseException is raised?

Reply:
OK, there must be some other issue then. The goal is to have OpenRefine return an error to the user rather than just stay frozen with its spinner up as shown by the screen capture and server traceback in issue #2903.